### DISCUSSION TITLE
0075 XLS-75d: Account Permission Delegation

### DISCUSSION THREAD

**Discussion by mvadari** [XRPLF/XRPL-Standards#218] [XLS Proposals]
## The earlier version can be found by looking at the edit history.

**Comment by tequdev** [comment]
It must be documented about Batch transaction.
1. all transaction types are available for Inner transactions if Batch transaction type is permissioned
OR
2. if Inner transaction transaction type is permissioned, Batch transaction is available
OR
3. both Batch and Inner transaction types must be permissioned

**Reply by mvadari** [reply]
That's a good point. I added a section on it in XLS-74d and an FAQ question here. The `OnBehalfOf` field may not be used on a `Batch` transaction, and must be used on the inner transactions instead.

**Comment by joshuahamsa** [comment]
The ability for an issuer to “blackhole” an account (not delegating payments) while still retaining operational abilities for everything else via another account would be great! (Even greater 3 years ago but 😅 🤷🏻‍♂️)

**Comment by dangell7** [comment]
Wouldn't a `TransactionTypeBitmask` be better than a list? Seems like just updating the bitmask would be much easier than getting the current list and checking the list for duplicates. Also for tooling it would be easier imo.

**Reply by mvadari** [reply]
A bitmask doesn't allow for granular permissions, which are where a lot of the value comes from. See [XLS-74d](https://github.com/XRPLF/XRPL-Standards/discussions/217).

It also functionally caps the total number of transactions in rippled at 256, though I think that is a less important concern.

**Comment by yinyiqian1** [comment]
regarding 3.1 fields, do we need `OwnerNode`?

**Reply by mvadari** [reply]
Not on the transaction, but yes on the ledger object - good catch, just updated that.

**Reply by yinyiqian1** [reply]
Yes. my typo, it should be 2.1 fields. Thanks for updating.

**Comment by yinyiqian1** [comment]
> `A permission may be delegated to as many accounts/signer lists as one is willing to pay reserve for.
> There is one reserve per delegated account.`

-----
Since we use `Account` and `Authorize` as key, the key will be something like: `keylet::accountPermission(account, authorize)`,
then the permission object on the ledger would be unique for a `Account` and `Authorize` pair.
If we wants to delegate the same permissions to two accounts, we need to create two permission objects even if sfPermissions field is the same. So a permission object can only be delegated to one account. And under this case, we can have the delegating account pay the reserve.

**Reply by mvadari** [reply]
Yes, that sounds correct.

**Comment by yinyiqian1** [comment]
Can we differentiate the transaction permission and granular permission in the json request by using `TransactionPermission` and `GranularPermission` field instead of `PermissionValue` so that we can parse these values easily, for example:
```
{
    "command": "submit",
    "tx_json" : {
      "TransactionType" : "AccountPermissionSet",
      "Account" : "rP5XKjUJk8S7fJs9F12YJFKGb8tvgwy5W",
      "Authorize" : "rGeQ2MufzFpdw8MEURxVx1DADi7BPbi9o3",
      "Permissions":[
        {"Permission": {"TransactionPermission": "TrustSet"}},
        {"Permission": {"TransactionPermission": "AMMCreate"}},
        {"Permission": {"GranularPermission": "TrustlineAuthorize"}}
      ]
   },
   "secret":"..."
}
```

**Reply by mvadari** [reply]
It doesn't feel very user-friendly to need different names depending on what type of value it is.

**Reply by yinyiqian1** [reply]
from code perspective, we are binding the field with the field type. You can refer to SFields.macro. It's better to define two types in the SFields.macro for Transaction level and Granular level permission object.
Otherwise, we need to use one type object of type UINT32 for both of them, and then do a cast in the STParsedJSON.cpp. To keep the code consistent and clean, it's better to have two type fields for them.

from user perspective, it's reasonable for them to have the knowledge that Permission type is different beforehand. These two object types have different field type (uint32 and uint16), it's normal that we give different field name for different object type in JSON.

**Reply by mvadari** [reply]
I think the `Permission` type _should_ be `UInt32` regardless of if it's granular or not. It'll also be easier to parse that way when you need to determine if someone has the right permissions to execute a transaction.

**Reply by ximinez** [reply]
@yinyiqian1 
> "secret":[....]

I hope this secret isn't used anywhere with real money. If it is, you need to rekey immediately.

**Reply by yinyiqian1** [reply]
No real money. just test in standalone mode. But I'll update that comment. Thanks for reminding.

**Comment by yinyiqian1** [comment]
Can we simplify the permission `Permissions: [{Permission: {PermissionValue: "TrustlineAuthorize"}}]` to `Permissions: [{Permission: "TrustlineAuthorize"}]`. for example:

```
"Permissions":[
        {"Permission": "TrustSet"},
        {"Permission": "AMMCreate"},
        {"Permission": "TrustlineAuthorize"}
      ]
```

**Reply by mvadari** [reply]
That's not possible. An STArray must contain STObjects, and an STObject must be represented as shown.

**Comment by yinyiqian1** [comment]
Should AccountPermissionSet transaction be banned from being delegated to other accounts?

**Reply by mvadari** [reply]
I would put it in the same category as `AccountSet`/`SignerListSet`/`SetRegularKey`.

**Comment by gregtatcam** [comment]
`Permissions` is an array of strings, which is pretty inefficient to store in the object. Should it be an array of `UINT32` instead? I understand that we want to have a user-friendly permissions but UI can taker care of this.

**Reply by mvadari** [reply]
The intent was for them to be stored as `UInt32`s and converted to strings in the JSON representation (like `TransactionType`).

**Reply by gregtatcam** [reply]
You meant to say `from strings`, right? So we are going to use it sort of like `enum` and we can pass a string in `UINT32`, which if `Permission` is converted to `UINT32`. It doesn't feel right having arbitrary strings parsed as `UINT32`. Maybe it's better to have its own `STI_PERMISSION` type instead?

**Reply by mvadari** [reply]
It's not arbitrary strings, it's very specific values - just like how `TransactionType` is displayed as a string in the JSON representation, but is encoded as a `UInt16` internally.

**Reply by gregtatcam** [reply]
`arbitrary` is wrong word. An `enum` name perhaps. `TransactionType` does set a precedent.

**Reply by mvadari** [reply]
I'm not opposed to adding a new SType for this, but it just seems like overkill to me.

**Comment by gregtatcam** [comment]
Shouldn't there be a transaction to delete `AccountPermission`?

**Reply by mvadari** [reply]
I was thinking an empty array would delete. If you'd rather have that be a separate transaction that's fine by me though.

**Reply by gregtatcam** [reply]
Either way is fine with me but I think we generally have a separate tx to delete.

**Comment by gregtatcam** [comment]
The first failure condition in `4.2. Failure Conditions` is `The OnBehalfOf account doesn't exist.` but `OnBehalfOf` is an optional field.

**Reply by yinyiqian1** [reply]
I think it means when the OnBehalfOf is given, and it does not exist. Then give the failure.

**Reply by mvadari** [reply]
Yes, exactly.

**Comment by yinyiqian1** [comment]
there should be an RPC call like `account_permission`, right? Shall we add it in the spec?

**Reply by mvadari** [reply]
Why is that necessary? Easy enough to just look it up via `ledger_entry`, right?

**Reply by yinyiqian1** [reply]
can we also add it in the spec as well so that the users know the parameters? for example:
```
{
  "method": "ledger_entry",
  "params" : [
    {
      "account_permission": {
      "account": "...",
      "authorize": "..."
    },
    "ledger_index": "validated"
    }
  ]
}
```

**Comment by mvadari** [comment]
Per the [XLS Contributing process](https://github.com/XRPLF/XRPL-Standards/blob/master/CONTRIBUTING.md), it is my opinion that we have reached a "well-refined standard."

As such, I propose that we move this discussion to a file (via https://github.com/XRPLF/XRPL-Standards/pull/257) and work on final changes using additional PRs, for better change-tracking.

Please comment here if you would like to object to moving this spec/discussion forward in the process into a DRAFT spec.

(Note that per the Contributing guidelines, moving a spec into the [DRAFT](https://github.com/XRPLF/XRPL-Standards/blob/master/CONTRIBUTING.md#drafts) state does not mean any kind of endorsement, nor does it mean that this specification will become adopted. It is solely meant as a mechanism to enable better change tracking using PRs.)

**Comment by ckeshava** [comment]
1. Where can I find more information about `sfDelegatingSeq` and `sfDelegatingTicketSeq` fields? 
2. Regarding the `Permissions` nested object, I am able to use the below snippet:
```
Permissions: [{Permission: {PermissionValue: 1+0 /* Proxy for the Payment transaction */}}]
```
This format works correctly and performs the expected functionality. In light of this observation, are both enum-type-strings and integers allowed in this nested object?

**Reply by yinyiqian1** [reply]
1.  These fields are added during implementation. To avoid double insertion for ledger object whose key is generated using account sequence, I added `sfDelegateSequence` as a common optional field to track delegating account's sequence. If A sends transaction on behalf of B, then both A and B's sequence are incremented.
Since a user can create ticket on behalf of another account,  so I added `sfDelegateTicketSequence` to track the ticket of the delegating account so that the delegated account can use a ticket he created on behalf of the delegating account.

2. Good catch. Previously, this is allowed for both string and integer. But I added a supported permission string list in the recent commit, could you use the updated code, this should fail.

**Reply by mvadari** [reply]
IMO only strings should be allowed.

**Comment by ckeshava** [comment]
What is the expected behavior of the `account_tx` RPC call? Suppose `alice` has delegated `Payment` transaction permissions to `bob`, what is the behavior of `account_tx(alice)` -- Will it return `bob`'s transaction where `sfOnBehalfOf == alice` ?

I didn't see any diff in `src/test/rpc/AccountTx_test.cpp` (or) `src/xrpld/rpc/handlers/AccountTx.cpp` in the implementation of this amendment.

@yinyiqian1

**Reply by yinyiqian1** [reply]
No. It won't show up in alice's account tx. It will be bob's tx.
@mvadari do you think we need to add it?

**Reply by mvadari** [reply]
I think it should show up in both automatically, in the same way that any IOU payment will show up in the issuer's `account_tx`.
